// Report sections
bm_sections = {};
bm_sections['deduplication'] = {{bm_deduplication}};
bm_sections['splitting'] = {{bm_splitting}};
bm_sections['nucleotide'] = {{bm_nucleotide}};
bm_sections['mbias'] = {{bm_mbias}};
bm_sections['mbias_2'] = {{bm_mbias_2}};

// All of the data
d = {};
d['unique_seqs'] = '{{unique_seqs}}';
d['no_alignments'] = '{{no_alignments}}';
d['multiple_alignments'] = '{{multiple_alignments}}';
d['no_genomic'] = '{{no_genomic}}';
d['number_OT'] = '{{number_OT}}';
d['number_CTOT'] = '{{number_CTOT}}';
d['number_CTOB'] = '{{number_CTOB}}';
d['number_OB'] = '{{number_OB}}';
d['number_OT'] = '{{number_OT}}';
d['number_CTOT'] = '{{number_CTOT}}';
d['number_CTOB'] = '{{number_CTOB}}';
d['number_OB'] = '{{number_OB}}';
d['number_OT'] = '{{number_OT}}';
d['number_CTOT'] = '{{number_CTOT}}';
d['number_CTOB'] = '{{number_CTOB}}';
d['number_OB'] = '{{number_OB}}';
d['number_OT'] = '{{number_OT}}';
d['number_CTOT'] = '{{number_CTOT}}';
d['number_CTOB'] = '{{number_CTOB}}';
d['number_OB'] = '{{number_OB}}';
d['number_OT'] = '{{number_OT}}';
d['number_CTOT'] = '{{number_CTOT}}';
d['number_CTOB'] = '{{number_CTOB}}';
d['number_OB'] = '{{number_OB}}';
d['perc_CpG_graph'] = '{{perc_CpG_graph}}';
d['perc_CHG_graph'] = '{{perc_CHG_graph}}';
d['perc_CHH_graph'] = '{{perc_CHH_graph}}';
d['unique_alignments_duplicates'] = '{{unique_alignments_duplicates}}';
d['duplicate_alignments_duplicates'] = '{{duplicate_alignments_duplicates}}';
d['perc_CpG_graph_splitting'] = '{{perc_CpG_graph_splitting}}';
d['perc_CHG_graph_splitting'] = '{{perc_CHG_graph_splitting}}';
d['perc_CHH_graph_splitting'] = '{{perc_CHH_graph_splitting}}';
d['nuc_A_p_exp'] = '{{nuc_A_p_exp}}';
d['nuc_T_p_exp'] = '{{nuc_T_p_exp}}';
d['nuc_C_p_exp'] = '{{nuc_C_p_exp}}';
d['nuc_G_p_exp'] = '{{nuc_G_p_exp}}';
d['nuc_AC_p_exp'] = '{{nuc_AC_p_exp}}';
d['nuc_CA_p_exp'] = '{{nuc_CA_p_exp}}';
d['nuc_TC_p_exp'] = '{{nuc_TC_p_exp}}';
d['nuc_CT_p_exp'] = '{{nuc_CT_p_exp}}';
d['nuc_CC_p_exp'] = '{{nuc_CC_p_exp}}';
d['nuc_CG_p_exp'] = '{{nuc_CG_p_exp}}';
d['nuc_GC_p_exp'] = '{{nuc_GC_p_exp}}';
d['nuc_GG_p_exp'] = '{{nuc_GG_p_exp}}';
d['nuc_AG_p_exp'] = '{{nuc_AG_p_exp}}';
d['nuc_GA_p_exp'] = '{{nuc_GA_p_exp}}';
d['nuc_TG_p_exp'] = '{{nuc_TG_p_exp}}';
d['nuc_GT_p_exp'] = '{{nuc_GT_p_exp}}';
d['nuc_TT_p_exp'] = '{{nuc_TT_p_exp}}';
d['nuc_TA_p_exp'] = '{{nuc_TA_p_exp}}';
d['nuc_AT_p_exp'] = '{{nuc_AT_p_exp}}';
d['nuc_AA_p_exp'] = '{{nuc_AA_p_exp}}';
d['nuc_A_p_obs'] = '{{nuc_A_p_obs}}';
d['nuc_T_p_obs'] = '{{nuc_T_p_obs}}';
d['nuc_C_p_obs'] = '{{nuc_C_p_obs}}';
d['nuc_G_p_obs'] = '{{nuc_G_p_obs}}';
d['nuc_AC_p_obs'] = '{{nuc_AC_p_obs}}';
d['nuc_CA_p_obs'] = '{{nuc_CA_p_obs}}';
d['nuc_TC_p_obs'] = '{{nuc_TC_p_obs}}';
d['nuc_CT_p_obs'] = '{{nuc_CT_p_obs}}';
d['nuc_CC_p_obs'] = '{{nuc_CC_p_obs}}';
d['nuc_CG_p_obs'] = '{{nuc_CG_p_obs}}';
d['nuc_GC_p_obs'] = '{{nuc_GC_p_obs}}';
d['nuc_GG_p_obs'] = '{{nuc_GG_p_obs}}';
d['nuc_AG_p_obs'] = '{{nuc_AG_p_obs}}';
d['nuc_GA_p_obs'] = '{{nuc_GA_p_obs}}';
d['nuc_TG_p_obs'] = '{{nuc_TG_p_obs}}';
d['nuc_GT_p_obs'] = '{{nuc_GT_p_obs}}';
d['nuc_TT_p_obs'] = '{{nuc_TT_p_obs}}';
d['nuc_TA_p_obs'] = '{{nuc_TA_p_obs}}';
d['nuc_AT_p_obs'] = '{{nuc_AT_p_obs}}';
d['nuc_AA_p_obs'] = '{{nuc_AA_p_obs}}';
d['CHH_total_calls_R1'] = '{{CHH_total_calls_R1}}';
d['CHG_total_calls_R1'] = '{{CHG_total_calls_R1}}';
d['CpG_total_calls_R1'] = '{{CpG_total_calls_R1}}';
d['CHH_methylation_R1'] = '{{CHH_methylation_R1}}';
d['CHG_methylation_R1'] = '{{CHG_methylation_R1}}';
d['CpG_methylation_R1'] = '{{CpG_methylation_R1}}';
d['CHH_total_calls_R2'] = '{{CHH_total_calls_R2}}';
d['CHG_total_calls_R2'] = '{{CHG_total_calls_R2}}';
d['CpG_total_calls_R2'] = '{{CpG_total_calls_R2}}';
d['CHH_methylation_R2'] = '{{CHH_methylation_R2}}';
d['CHG_methylation_R2'] = '{{CHG_methylation_R2}}';
d['CpG_methylation_R2'] = '{{CpG_methylation_R2}}';

// Convert strings to numbers
for (k in d){
  if(!isNaN(d[k])){
    d[k] = parseFloat(d[k]);
  }
}

// Alignment
$(function () {
  
  // Set defaults for Highcharts
  Highcharts.setOptions({
    colors: ['#0d233a','#2f7ed8','#8bbc21','#910000','#1aadce','#492970','#f28f43','#77a1e5','#c42525','#a6c96a'],
    title: { text: null },
    credits: { enabled: false }
  });
  
  // Alignment
  $('#alignment').highcharts({
    chart: { type: 'pie' },
    tooltip: {
      formatter: function() {
        return '<b>'+ this.point.name + ':</b>'+this.percentage.toFixed(1) + '%';
      }
    },
    plotOptions: {
      pie: {
        allowPointSelect: true,
        cursor: 'pointer',
        dataLabels: {
          enabled: true,
          color: '#000000',
          connectorColor: '#000000',
          format: '<b>{point.name}</b>: {point.percentage:.1f} %'
        }
      }
    },
    series: [{
      type: 'pie',
      name: 'Reads',
      data: [
        {
          name: 'Unique Alignments',
          y: d['unique_seqs'],
          sliced: true,
          selected: true
        },
        ['No Alignment', d['no_alignments']],
        ['Multiple Alignments', d['multiple_alignments']],
        ['No Genomic Sequence', d['no_genomic']]
      ]
    }]
  });
  
  // Methylation Context
  $('#methylation_context').highcharts({
    chart: { type: 'column' },
    xAxis: { categories: ['CpG', 'CHG', 'CHH'] },
    yAxis: {
      title: { text: '% Methylation' },
      max: 100
    },
    legend: { enabled: false },
    tooltip: {
      headerFormat: '',
      pointFormat: '<b>{point.category} Methylation</b>: {point.y}%'
    },
    series: [{
      name: 'Methylation Context',
      data: [d['perc_CpG_graph'], d['perc_CHG_graph'], d['perc_CHH_graph']]
    }]
  });
  
  // Aligment Context
  $('#alignment_context').highcharts({
    chart: { type: 'column' },
    xAxis: { categories: ['OT', 'CTOT', 'CTOB', 'OB' ] },
    yAxis: { title: { text: 'Number of Alignments' } },
    tooltip: {
      formatter: function() {
        if(this.x == 'OT'){
          return '<b>' + this.x + ':</b> ' + (this.y / ( d['number_OT'] + d['number_CTOT'] + d['number_CTOB'] + d['number_OB'] ) * 100).toFixed(1) + '%'
          + '<br><em>original top strand</em>';
        } else if(this.x == 'CTOT'){
          return '<b>' + this.x + ':</b> ' + (this.y / ( d['number_OT'] + d['number_CTOT'] + d['number_CTOB'] + d['number_OB'] ) * 100).toFixed(1) + '%'
          + '<br><em>complementary to original top strand</em>';
        } else if(this.x == 'CTOB'){
          return '<b>' + this.x + ':</b> ' + (this.y / ( d['number_OT'] + d['number_CTOT'] + d['number_CTOB'] + d['number_OB'] ) * 100).toFixed(1) + '%'
          + '<br><em>complementary to original bottom strand</em>';
        } else if(this.x == 'OB'){
          return '<b>' + this.x + ':</b> ' + (this.y / ( d['number_OT'] + d['number_CTOT'] + d['number_CTOB'] + d['number_OB'] ) * 100).toFixed(1) + '%'
          + '<br><em>original bottom strand</em>';
        } else {
          return false;
        }
      }
    },
    series: [{
      name: 'Alignment Context',
      data: [ d['number_OT'], d['number_CTOT'], d['number_CTOB'], d['number_OB'] ]
    }]
  });
  
  // Deduplication
  if(bm_sections['deduplication']){
    $('#bm_deduplication').show();
    $('#deduplication_plot').highcharts({
      chart: { type: 'pie' },
      tooltip: {
        formatter: function() {
          return '<b>'+ this.point.name + ':</b>'+this.percentage.toFixed(1) + '%';
        }
      },
      plotOptions: {
        pie: {
          allowPointSelect: true,
          cursor: 'pointer',
          dataLabels: {
            enabled: true,
            color: '#000000',
            connectorColor: '#000000',
            format: '<b>{point.name}</b>: {point.percentage:.1f} %'
          }
        }
      },
      series: [{
        type: 'pie',
        name: 'Read Pairs',
        data: [
          ['Unique', d['unique_alignments_duplicates']],
          ['Duplicate', d['duplicate_alignments_duplicates']]
        ]
      }]
    });
  }
  
  if(bm_sections['splitting']){
    $('#bm_splitting').show();
    // Methylation Context after Extraction
    $('#dedup_methylation_context').highcharts({
      chart: { type: 'column' },
      xAxis: { categories: ['CpG', 'CHG', 'CHH'] },
      yAxis: {
        title: { text: '% Methylation' },
        max: 100
      },
      legend: { enabled: false },
      tooltip: {
        headerFormat: '',
        pointFormat: '<b>{point.category} Methylation</b>: {point.y}%'
      },
      series: [{
        name: 'Methylation Context',
        data: [d['perc_CpG_graph_splitting'], d['perc_CHG_graph_splitting'], d['perc_CHH_graph_splitting']]
      }]
    });
  }
  
  // Nucleotide coverage
  if(bm_sections['nucleotide']){
    $('#bm_nucleotide').show();
    $('#nucleotide_coverage').highcharts({
      chart: {
        type: 'bar'
      },
      xAxis: {
        categories: ['A','T','C','G','AC','CA','TC','CT','CC','CG','GC','GG','AG','GA','TG','GT','TT','TA','AT','AA'],
        title: { text: null }
      },
      yAxis: {
        min: 0,
        title: { text: null }
      },
      tooltip: {
        valueDecimals: 1,
        valueSuffix: '%',
        shared: true,
        headerFormat: '<span style="text-decoration:underline; font-weight:bold; font-size: 1.1em;">{point.key}</span><br/>'
      },
      plotOptions: {
        series: {
          borderWidth: 0,
          groupPadding: 0.1
        }
      },
      series: [{
        name: 'Percent Genomic',
        data: [
          d['nuc_A_p_exp'],d['nuc_T_p_exp'],d['nuc_C_p_exp'],d['nuc_G_p_exp'],
          d['nuc_AC_p_exp'],d['nuc_CA_p_exp'],d['nuc_TC_p_exp'],d['nuc_CT_p_exp'],
          d['nuc_CC_p_exp'],d['nuc_CG_p_exp'],d['nuc_GC_p_exp'],d['nuc_GG_p_exp'],
          d['nuc_AG_p_exp'],d['nuc_GA_p_exp'],d['nuc_TG_p_exp'],d['nuc_GT_p_exp'],
          d['nuc_TT_p_exp'],d['nuc_TA_p_exp'],d['nuc_AT_p_exp'],d['nuc_AA_p_exp'],
        ]
      }, {
        name: 'Percent Sample',
        data: [
          d['nuc_A_p_obs'],d['nuc_T_p_obs'],d['nuc_C_p_obs'],d['nuc_G_p_obs'],
          d['nuc_AC_p_obs'],d['nuc_CA_p_obs'],d['nuc_TC_p_obs'],d['nuc_CT_p_obs'],
          d['nuc_CC_p_obs'],d['nuc_CG_p_obs'],d['nuc_GC_p_obs'],d['nuc_GG_p_obs'],
          d['nuc_AG_p_obs'],d['nuc_GA_p_obs'],d['nuc_TG_p_obs'],d['nuc_GT_p_obs'],
          d['nuc_TT_p_obs'],d['nuc_TA_p_obs'],d['nuc_AT_p_obs'],d['nuc_AA_p_obs'],
        ]
      }]
    });
  }

  // M-Bias Plots
  if(bm_sections['mbias']){
    $('#bm_mbias').show();
    // Set plot defaults to be used twice
    var m_bias_settings = {
      colors: [ '#CCF0E1','#EDD3A8','#69798A','#21BCA2','#F29D13','#0d233a','#f28f43','#77a1e5','#c42525','#a6c96a'],
      chart: {
        zoomType: 'x',
        marginRight: 80,
        alignTicks: false,
        plotBorderWidth:1
      },
      title: { text: 'Read 1' },
      subtitle: {
        text: document.ontouchstart === undefined ?
        'Click and drag in the plot area to zoom in' :
        'Drag your finger over the plot to zoom in'
      },
      xAxis: {
        title: { text: 'Position (bp)' },
        tickInterval: 5,
        minTickInterval: 0,
        min: 0,
      },
      yAxis: [{
        title: { text: '% Methylation' },
        min: 0,
        max: 100,
        gridLineWidth: 0,
        tickWidth: 1,
        lineWidth: 1
      }, { // secondary axis
        title: { text: '# Methylation Calls' },
        min:0,
        gridLineWidth: 0,
        lineWidth: 1,
        tickWidth: 1,
        opposite: true
      }],
      credits: { enabled: false },
      tooltip: {
        shared: true,
        headerFormat: '<b>Base {point.x}</b><table>',
        pointFormat: '<tr><td><span style="color:{series.color};">{series.name} Methylation:</span></td><td>{point.y}</td></tr>',
        footerFormat: '</table>',
        useHTML: true
      },
      legend: {
        layout: 'vertical',
        align: 'right',
        verticalAlign: 'top',
        x: -60,
        backgroundColor: '#FFFFFF'
      },
      plotOptions: {
        line: {
          lineWidth: 1,
          marker: { enabled: false },
          shadow: false
        }
      },
      series: [{
        name: 'CHH Total Calls',
        yAxis: 1,
        lineWidth: 2,
        legendIndex: 5,
        data: d['CHH_total_calls_R1']
      },{
        name: 'CHG Total Calls',
        yAxis: 1,
        lineWidth: 2,
        legendIndex: 4,
        data: d['CHG_total_calls_R1']
      },{
        name: 'CpG Total Calls',
        yAxis: 1,
        lineWidth: 2,
        legendIndex: 3,
        data: d['CpG_total_calls_R1']
      },{
        name: 'CHH Methylation',
        yAxis: 0,
        lineWidth: 3,
        legendIndex: 2,
        data: d['CHH_methylation_R1']
      },{
        name: 'CHG Methylation',
        yAxis: 0,
        lineWidth: 3,
        legendIndex: 1,
        data: d['CHG_methylation_R1']
      },{
        name: 'CpG Methylation',
        yAxis: 0,
        lineWidth: 3,
        legendIndex: 0,
        data: d['CpG_methylation_R1']
      }]
    };
    // Plot the graph
    $('#m_bias_1').highcharts(m_bias_settings);
    
    if(bm_sections['mbias_2']){
      $('#m_bias_2').show();
      // M-Bias Plot 2
      var m_bias_settings = {
        colors: [ '#CCF0E1','#EDD3A8','#69798A','#21BCA2','#F29D13','#0d233a','#f28f43','#77a1e5','#c42525','#a6c96a'],
        chart: {
          zoomType: 'x',
          marginRight: 80,
          alignTicks: false,
          plotBorderWidth:1
        },
        subtitle: {
          text: document.ontouchstart === undefined ?
          'Click and drag in the plot area to zoom in' :
          'Drag your finger over the plot to zoom in'
        },
        xAxis: {
          title: { text: 'Position (bp)' },
          tickInterval: 5,
          minTickInterval: 0,
          min: 0,
        },
        yAxis: [{
          title: { text: '% Methylation' },
          min: 0,
          max: 100,
          gridLineWidth: 0,
          tickWidth: 1,
          lineWidth: 1
        }, { // secondary axis
          title: { text: '# Methylation Calls' },
          min:0,
          gridLineWidth: 0,
          lineWidth: 1,
          tickWidth: 1,
          opposite: true
        }],
        credits: { enabled: false },
        tooltip: {
          shared: true,
          headerFormat: '<b>Base {point.x}</b><table>',
          pointFormat: '<tr><td><span style="color:{series.color};">{series.name} Methylation:</span></td><td>{point.y}</td></tr>',
          footerFormat: '</table>',
          useHTML: true
        },
        legend: {
          layout: 'vertical',
          align: 'right',
          verticalAlign: 'top',
          x: -60,
          backgroundColor: '#FFFFFF'
        },
        plotOptions: {
          line: {
            lineWidth: 1,
            marker: { enabled: false },
            shadow: false
          }
        },
        title: {
          text: 'Read 2'
        },
        series: [{
          name: 'CHH Total Calls',
          yAxis: 1,
          lineWidth: 2,
          legendIndex: 5,
          data: d['CHH_total_calls_R2']
        },{
          name: 'CHG Total Calls',
          yAxis: 1,
          lineWidth: 2,
          legendIndex: 4,
          data: d['CHG_total_calls_R2']
        },{
          name: 'CpG Total Calls',
          yAxis: 1,
          lineWidth: 2,
          legendIndex: 3,
          data: d['CpG_total_calls_R2']
        },{
          name: 'CHH Methylation',
          yAxis: 0,
          lineWidth: 3,
          legendIndex: 2,
          data: d['CHH_methylation_R2']
        },{
          name: 'CHG Methylation',
          yAxis: 0,
          lineWidth: 3,
          legendIndex: 1,
          data: d['CHG_methylation_R2']
        },{
          name: 'CpG Methylation',
          yAxis: 0,
          lineWidth: 3,
          legendIndex: 0,
          data: d['CpG_methylation_R2']
        }]
      }

      // Plot graph
      $('#m_bias_2').highcharts(m_bias_settings);
    }
  }


  // adds commas to big numbers for better readability
  $('table tbody tr td').each(function(){
    if(isNumber($(this).text())){
      $(this).text(addCommas($(this).text()));
    }
  });
  function isNumber(n) { return !isNaN(parseFloat(n)) && isFinite(n); }
  function addCommas(nStr) {
    nStr += '';
    x = nStr.split('.');
    x1 = x[0];
    x2 = x.length > 1 ? '.' + x[1] : '';
    var rgx = /(\d+)(\d{3})/;
    while (rgx.test(x1)) {
      x1 = x1.replace(rgx, '$1' + ',' + '$2');
    }
    return x1 + x2;
  }
});